packages/databus/src/dataregistry.ts

```diff
@@ -0,0 +1,136 @@
+/*-----------------------------------------------------------------------------
+| Copyright (c) Jupyter Development Team.
+| Distributed under the terms of the Modified BSD License.
+|----------------------------------------------------------------------------*/
+
+import { DisposableDelegate, IDisposable } from '@phosphor/disposable';
+
+import { ISignal, Signal } from '@phosphor/signaling';
+
+/**
+ * An interface for an abstract dataset, with a mimetype, URI, and data.
+ *
+ * This is templated over the type of data `T`.
+ *
+ * #### Notes
+ * This interface is similar conceptually to that of nbformat.IMimeBundle,
+ * but with only a single mimetype and different structure. We expect there
+ * to be utilities to convert between this formats.
+ */
+export interface IDataset<T> {
+  /**
+   * The string mimetype for the dataset.
+   */
+  mimeType: string;
+
+  /**
+   * A persistent URI that points to the dataset.
+   *
+   * #### Notes
+   * This can be used by other extensions and services to maintain
+   * persistent pointers to datasets across sessions. This is needed
+   * as the datasets themselves are not assumed to be serializable.
+   */
+  uri?: string;
+
+  /***
+   * The data of the dataset.
+   *
+   * #### Notes
+   * The notion of data in the databus is completely abstract, and templated
+   * over. Publishers and consumers of a given mimetype will need to agree
+   * what the actual data is and cast it appropriately before use.
+   */
+  data: T;
+}
+
+/**
+ * A data registry object, for managing datasets in JupyterLab.
+ */
+export class DataRegistry {
+  /**
+   * Construct a new data registry.
+   */
+  constructor() {}
+
+  /**
+   * Publish a dataset to the data registry.
+   *
+   * @param dataset - the `IDataset` to publish to the data registry.
+   *
+   * @returns A disposable which will remove the dataset.
+   *
+   * @throws An error if the given dataset is already published.
+   */
+  publish(dataset: IDataset<any>): IDisposable {
+    if (this._datasets.has(dataset)) {
+      throw new Error(`Dataset already published`);
+    }
+
+    this._datasets.add(dataset);
+
+    this._datasetsChanged.emit({ dataset, type: 'added' });
+
+    return new DisposableDelegate(() => {
+      this._datasets.delete(dataset);
+      this._datasetsChanged.emit({ dataset, type: 'removed' });
+    });
+  }
+
+  /**
+   * Filter the published datasets using a filtering function.
+   *
+   * @param func - A function to use for filtering.
+   *
+   * @returns An set of matching `IDataset` objects.
+   */
+  filter<T>(func: (value: IDataset<T>) => boolean): Set<IDataset<T>> {
```

Comment:
```suggestion
  filter<T extends IDataset<any>>(func: (value: IDataset<any>) => value is T): Set<T> {
```

Reply:
These suggestions change the filter function to take a [user defined type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards) instead of just a function that returns a boolean. This allows us to skip the `as IDataset<T>` later on, because Typescript knows that if this function returns `True`, then `value` is of type  `T`.